### PR TITLE
Refactor and extend RF interface for more flexible kernels. 

### LINKED ISSUE #228: interface for defining user-priors
Specifically, provide a modular interface that defines different types of covariance structures within a prior for the random feature hyperparameter learning.

### LINKED ISSUE #229: SVD truncation for vector RF

### LINKED ISSUE #230: Sample stage for GCM application

### PR DESCRIPTION
**Pull request by odunbar**
## Purpose 
Addresses  points in #215 In particular:
- Closes #228 
- Closes #229 
- Closes #230

## Content

- Adds MCMC stage into `examples/GCM` and improves running interface
- Adds Shrinkage estimation into cross-validation for RF hyperparameter optimization.
- Adds Kernel interface for RF, where one specifies either a `Separable` or `Nonseparable` Kernel followed by the structure of the covariance to be used to sample the features.
    - `Nonseparable(cov)`: define one covariance structure on the `p * d`-dim space,
    -  `Separable(in_cov,out_cov)`: define one `d`-dim covariance for inputs and one `p`-dim covariance for outputs
- Choose covariance structures for `cov`,`in_cov`,`out_cov` from 
    - `OneDimFactor()`: Must be selected for 1D covariances
    - `DiagonalFactor(eps)`: Diagonal covariance plus `eps*I` 
    - `CholeskyFactor(eps)`: Cholesky representation `L*L^T` plus eps*I` 
    - `LowRankFactor(rank, eps)`: Symmetric factorization `W*W^T` where `W = 1+UDU'` D diag, U rectangular, plus `eps*I`
    - `HierarchicalLowRankFactor(rank, eps)`: Symmetric factorization `W*W^T` where `W = 1+UXU'` U rectangular, plus `eps*I` X cholesky factored
- Rebalancing the loss function seems to get far more robust results for vector RF
 

----
- [x] I have read and checked the items on the review checklist.

### CODE REVIEW

**Comment by odunbar** [comment]
bors try
